operator/src/main/java/org/keycloak/operator/controllers/KeycloakDiscoveryServiceDependentResource.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2021 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.keycloak.operator.controllers;
+
+import io.fabric8.kubernetes.api.model.Service;
+import io.fabric8.kubernetes.api.model.ServiceBuilder;
+import io.fabric8.kubernetes.api.model.ServiceSpec;
+import io.fabric8.kubernetes.api.model.ServiceSpecBuilder;
+import io.javaoperatorsdk.operator.api.reconciler.Context;
+import io.javaoperatorsdk.operator.api.reconciler.ResourceDiscriminator;
+import io.javaoperatorsdk.operator.processing.dependent.kubernetes.CRUDKubernetesDependentResource;
+import io.javaoperatorsdk.operator.processing.dependent.kubernetes.KubernetesDependent;
+import io.javaoperatorsdk.operator.processing.event.ResourceID;
+import io.javaoperatorsdk.operator.processing.event.source.informer.InformerEventSource;
+
+import org.keycloak.operator.Constants;
+import org.keycloak.operator.crds.v2alpha1.deployment.Keycloak;
+
+import java.util.Optional;
+
+@KubernetesDependent(labelSelector = Constants.DEFAULT_LABELS_AS_STRING, resourceDiscriminator = KeycloakDiscoveryServiceDependentResource.NameResourceDiscriminator.class)
+public class KeycloakDiscoveryServiceDependentResource extends CRUDKubernetesDependentResource<Service, Keycloak> {
+
+    public static class NameResourceDiscriminator implements ResourceDiscriminator<Service, Keycloak> {
+        @Override
+        public Optional<Service> distinguish(Class<Service> resource, Keycloak primary, Context<Keycloak> context) {
+            InformerEventSource<Service, Keycloak> ies = (InformerEventSource<Service, Keycloak>) context
+                    .eventSourceRetriever().getResourceEventSourceFor(Service.class);
+
+            return ies.get(new ResourceID(getName(primary), primary.getMetadata().getNamespace()));
+        }
+    }
```

Comment:
Can we make this more generic (as it's shared between our services), or perhaps use [ResourceIDMatcherDiscriminator](https://github.com/operator-framework/java-operator-sdk/blob/b66ef938a9290a1dbf83b35e152f144cc01bcb8a/operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/reconciler/ResourceIDMatcherDiscriminator.java)?

Reply:
You could, but it encapsulates only a small amount of logic that is used in two places and ends up being more total code.  I'd settle for just factoring out a static method.

Reply:
I assumed we could do something as simple as
```java
    public static class NameResourceDiscriminator extends ResourceIDMatcherDiscriminator<Service, Keycloak> {
        public NameResourceDiscriminator() {
            super((kc) -> new ResourceID(getName(kc), kc.getMetadata().getNamespace()));
        }
    }
```
but a static method probably works too ;)

Reply:
On the other hand it feels a bit weird to call `KeycloakServiceDependentResource.getService` from 
 `KeycloakDiscoveryServiceDependentResource`. It's two separate services after all.

But it's not a blocker for me.

Reply:
> I assumed we could do something as simple as

What I don't like about that is it's based upon list iteration, rather than a name lookup.  Granted there shouldn't be many instances - it's just odd that the name is not a primary construct here.  I'd say that this is more of a case where something is missing from the JOSDK or at least being made more complicated.

Reply:
Sounds like an RFE to JOSDK then ;)

Reply:
https://github.com/operator-framework/java-operator-sdk/issues/2011